File: apps/web/lib/server/huggingface.ts
import type { AnalysisArtifacts, TimelineSegment } from "../types";
import { calibrationVersion, roundScore, type DetectorOutput, type VideoFrameInput } from "./detectors";

export const defaultTextModel = "desklib/ai-text-detector-v1.01";
export const defaultTextFallbackModels = ["Oxidane/tmr-ai-text-detector"];
export const defaultImageModel = "haywoodsloan/ai-image-detector-dev-deploy";
export const defaultImageFallbackModels = ["Ateeqq/ai-vs-human-image-detector", "umm-maybe/AI-image-detector"];
export const defaultVideoFrameModel = "haywoodsloan/ai-image-detector-dev-deploy";
export const defaultEndpointBase = "https://router.huggingface.co/hf-inference/models";

const timeoutMs = 12000;

export interface HfClassification {
  label: string;
  score: number;
}

export interface HfAttempt {
  model: string;
  status: "ok" | "error";
  label?: string;
  score?: number;
  error?: string;
}

interface HfScore {
  label: string;
  model: string;
  score: number;
  attempts: HfAttempt[];
}

export class ModelUnavailableError extends Error {
  constructor(message = "Model inference is unavailable. Configure HF_TOKEN or try again after the model provider recovers.") {
    super(message);
    this.name = "ModelUnavailableError";
  }
}

function metadataFor(artifacts: AnalysisArtifacts) {
  artifacts.metadata ??= {};
  return artifacts.metadata;
}

function envList(name: string, fallback: string[]) {
  const raw = process.env[name];
  if (!raw) return fallback;
  const values = raw
    .split(",")
    .map((entry) => entry.trim())
    .filter(Boolean);
  return values.length ? values : fallback;
}

export function configuredTextModels() {
  return [process.env.HF_TEXT_MODEL ?? defaultTextModel, ...envList("HF_TEXT_FALLBACK_MODELS", defaultTextFallbackModels)];
}

export function configuredImageModels() {
  return [process.env.HF_IMAGE_MODEL ?? defaultImageModel, ...envList("HF_IMAGE_FALLBACK_MODELS", defaultImageFallbackModels)];
}

export function configuredVideoFrameModels() {
  const primary = process.env.HF_VIDEO_FRAME_MODEL ?? defaultVideoFrameModel;
  return [primary, ...envList("HF_IMAGE_FALLBACK_MODELS", defaultImageFallbackModels)];
}

export function hasHfToken() {
  return Boolean(process.env.HF_TOKEN);
}

function strictModelMode() {
  return process.env.TRUTHLENS_REQUIRE_HF === "true" || process.env.VERCEL_ENV === "production";
}

function errorMessage(caught: unknown) {
  return caught instanceof Error ? caught.message.slice(0, 180) : "HF inference failed";
}

export function flattenResponse(body: unknown): HfClassification[] {
  const value = Array.isArray(body) && Array.isArray(body[0]) ? body[0] : body;
  if (!Array.isArray(value)) {
    if (
      typeof value === "object" &&
      value !== null &&
      "label" in value &&
      "score" in value &&
      typeof value.label === "string" &&
      typeof value.score === "number"
    ) {
      return [{ label: value.label, score: value.score }];
    }
    return [];
  }
  return value
    .filter((entry): entry is HfClassification => {
      return (
        typeof entry === "object" &&
        entry !== null &&
        "label" in entry &&
        "score" in entry &&
        typeof entry.label === "string" &&
        typeof entry.score === "number"
      );
    })
    .sort((first, second) => second.score - first.score);
}

function normalizeLabel(label: string) {
  return label.toLowerCase().replace(/[_-]+/g, " ").trim();
}

export function isAiLabel(label: string) {
  const normalized = normalizeLabel(label);
  if (/^(label 1|ai|artificial|fake|generated|synthetic|machine)$/.test(normalized)) return true;
  if (/(not ai|not generated|non ai|human|real|authentic|original)/.test(normalized)) return false;
  return /(ai|generated|machine|synthetic|fake|artificial)/.test(normalized);
}

export function isHumanLabel(label: string) {
  const normalized = normalizeLabel(label);
  if (/^(label 0|human|real|authentic|original)$/.test(normalized)) return true;
  return /(not ai|not generated|non ai|human|real|authentic|original)/.test(normalized);
}

export function aiScoreFromLabels(labels: HfClassification[]) {
  const ai = labels.find((entry) => isAiLabel(entry.label));
  if (ai) return { label: ai.label, score: roundScore(ai.score) };
  const human = labels.find((entry) => isHumanLabel(entry.label));
  if (human) return { label: human.label, score: roundScore(1 - human.score) };
  return null;
}

async function classifyWithHuggingFace(input: string | Buffer, model: string, token: string): Promise<Omit<HfScore, "attempts">> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), timeoutMs);
  const isBuffer = Buffer.isBuffer(input);
  try {
    const response = await fetch(`${process.env.HF_INFERENCE_ENDPOINT_BASE ?? defaultEndpointBase}/${model}`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${token}`,
        ...(isBuffer ? { "Content-Type": "application/octet-stream" } : { "Content-Type": "application/json" })
      },
      body: isBuffer ? input : JSON.stringify({ inputs: input, parameters: { top_k: 2, function_to_apply: "softmax" } }),
      signal: controller.signal
    });
    if (!response.ok) {
      const body = await response.text().catch(() => "");
      throw new Error(`HF inference returned ${response.status}${body ? `: ${body.slice(0, 120)}` : ""}`);
    }
    const labels = flattenResponse(await response.json());
    const mapped = aiScoreFromLabels(labels);
    if (!mapped) {
      throw new Error("HF inference returned no mappable AI/human label");
    }
    return { model, ...mapped };
  } finally {
    clearTimeout(timeout);
  }
}

export async function classifyWithFallbacks(input: string | Buffer, models: string[], token: string): Promise<HfScore> {
  const attempts: HfAttempt[] = [];
  const uniqueModels = Array.from(new Set(models));
  for (const model of uniqueModels) {
    try {
      const score = await classifyWithHuggingFace(input, model, token);
      attempts.push({ model, status: "ok", label: score.label, score: score.score });
      return { ...score, attempts };
    } catch (caught) {
      attempts.push({ model, status: "error", error: errorMessage(caught) });
    }
  }
  const lastError = attempts.at(-1)?.error ?? "All HF model attempts failed.";
  throw Object.assign(new ModelUnavailableError(lastError), { attempts });
}

function applyHfNeuralScore(output: DetectorOutput, score: HfScore, explanation: string) {
  const metadata = metadataFor(output.artifacts);
  output.artifacts.metadata = {
    ...metadata,
    hf_model: score.model,
    hf_label: score.label,
    hf_score: score.score,
    hf_status: "ok",
    model_attempts: score.attempts,
    calibration_version: calibrationVersion
  };
  output.layer_scores.neural_classifier = score.score;
  output.layer_breakdown = output.layer_breakdown.map((layer) =>
    layer.name === "Neural Classifier" ? { ...layer, score: score.score, explanation } : layer
  );
}

function applyHfError(output: DetectorOutput, model: string, caught: unknown) {
  const attempts = typeof caught === "object" && caught !== null && "attempts" in caught ? caught.attempts : [{ model, status: "error", error: errorMessage(caught) }];
  output.artifacts.metadata = {
    ...metadataFor(output.artifacts),
    hf_model: model,
    hf_status: "fallback",
    hf_error: errorMessage(caught),
    model_attempts: attempts,
    calibration_version: calibrationVersion
  };
}

function unavailableIfStrict(output: DetectorOutput, model: string, caught: unknown) {
  applyHfError(output, model, caught);
  if (strictModelMode()) {
    throw caught instanceof ModelUnavailableError ? caught : new ModelUnavailableError(errorMessage(caught));
  }
}

function applyVideoFrameScores(output: DetectorOutput, frameScores: HfScore[], frameInputs: VideoFrameInput[]) {
  const scores = frameScores.map((entry) => entry.score);
  const average = scores.reduce((total, value) => total + value, 0) / scores.length;
  const max = Math.max(...scores);
  const aggregate = roundScore(average * 0.65 + max * 0.35);
  const attempts = frameScores.flatMap((entry, index) =>
    entry.attempts.map((attempt) => ({
      ...attempt,
      frame_index: frameInputs[index]?.index ?? index,
      frame_seconds: frameInputs[index]?.seconds ?? index
    }))
  );
  const timeline: TimelineSegment[] = frameScores.map((score, index) => ({
    start_seconds: frameInputs[index]?.seconds ?? index,
    end_seconds: frameInputs[index + 1]?.seconds ?? (frameInputs[index]?.seconds ?? index) + 1,
    score: score.score
  }));

  output.artifacts.timeline = timeline;
  output.artifacts.metadata = {
    ...metadataFor(output.artifacts),
    hf_model: frameScores[0]?.model ?? configuredVideoFrameModels()[0],
    hf_label: "frame_aggregate",
    hf_score: aggregate,
    hf_status: "ok",
    sampled_frame_count: frameScores.length,
    model_attempts: attempts,
    calibration_version: calibrationVersion
  };
  output.layer_scores.neural_classifier = aggregate;
  output.layer_scores.temporal_forensics = roundScore(average);
  output.layer_scores.face_mesh = roundScore(max);
  output.layer_breakdown = output.layer_breakdown.map((layer) => {
    if (layer.name === "Neural Classifier") {
      return { ...layer, score: aggregate, explanation: "Aggregate Hugging Face image-detector score across sampled video frames." };
    }
    if (layer.name === "Temporal Forensics") {
      return { ...layer, score: roundScore(average), explanation: "Average sampled-frame AI probability." };
    }
    if (layer.name === "Face Mesh") {
      return { ...layer, score: roundScore(max), explanation: "Highest sampled-frame AI probability, useful for localized synthetic edits." };
    }
    return layer;
  });
  output.detected_sources =
    aggregate >= 0.58 ? ["AI-generated video frames", "Hugging Face image detector"] : ["No specific generator identified"];
}

export async function enrichWithHuggingFace(output: DetectorOutput, input: { text?: string; image?: Buffer; videoFrames?: VideoFrameInput[] }) {
  const token = process.env.HF_TOKEN;
  if (!token) {
    output.artifacts.metadata = {
      ...metadataFor(output.artifacts),
      hf_status: "disabled",
      calibration_version: calibrationVersion
    };
    if (strictModelMode()) {
      throw new ModelUnavailableError("HF_TOKEN is required for production model-backed analysis.");
    }
    return output;
  }

  if (output.modality === "TEXT" && input.text) {
    const models = configuredTextModels();
    try {
      const score = await classifyWithFallbacks(input.text, models, token);
      applyHfNeuralScore(output, score, "Hugging Face AI text detector probability. Final confidence is calibrated from this model score.");
      output.explanation_parts.push("The neural classifier layer used Hugging Face text detector inference.");
    } catch (caught) {
      unavailableIfStrict(output, models[0], caught);
    }
  }

  if (output.modality === "IMAGE" && input.image) {
    const models = configuredImageModels();
    try {
      const score = await classifyWithFallbacks(input.image, models, token);
      applyHfNeuralScore(output, score, "Hugging Face AI image detector probability. Final confidence is calibrated from this model score.");
      output.explanation_parts.push("The neural classifier layer used Hugging Face image detector inference.");
      output.detected_sources = score.score >= 0.58 ? ["Hugging Face AI image detector"] : ["No specific generator identified"];
    } catch (caught) {
      unavailableIfStrict(output, models[0], caught);
    }
  }

  if (output.modality === "VIDEO" && input.videoFrames?.length) {
    const models = configuredVideoFrameModels();
    try {
      const frameScores = await Promise.all(input.videoFrames.slice(0, 6).map((frame) => classifyWithFallbacks(frame.data, models, token)));
      applyVideoFrameScores(output, frameScores, input.videoFrames.slice(0, 6));
      output.explanation_parts.push("The video layer used Hugging Face image detector inference over sampled frames.");
    } catch (caught) {
      unavailableIfStrict(output, models[0], caught);
    }
  }

  return output;
}

File: apps/api/app/huggingface.py
from __future__ import annotations

import json
import re
import urllib.error
import urllib.request
from typing import Any

from app.config import Settings
from app.detectors import DetectorOutput, round_score


TIMEOUT_SECONDS = 8


def _flatten_response(body: Any) -> list[dict[str, Any]]:
    value = body[0] if isinstance(body, list) and body and isinstance(body[0], list) else body
    if not isinstance(value, list):
        return []
    labels = [
        entry
        for entry in value
        if isinstance(entry, dict) and isinstance(entry.get("label"), str) and isinstance(entry.get("score"), (int, float))
    ]
    return sorted(labels, key=lambda entry: float(entry["score"]), reverse=True)


def _normalized_label(label: str) -> str:
    return re.sub(r"[_-]+", " ", label.lower()).strip()


def _is_ai_label(label: str) -> bool:
    normalized = _normalized_label(label)
    if re.fullmatch(r"(label 1|ai|artificial|fake|generated|synthetic|machine)", normalized):
        return True
    if re.search(r"(not ai|not generated|non ai|human|real|authentic|original)", normalized):
        return False
    return bool(re.search(r"(ai|generated|machine|synthetic|fake|artificial)", normalized))


def _is_human_label(label: str) -> bool:
    normalized = _normalized_label(label)
    if re.fullmatch(r"(label 0|human|real|authentic|original)", normalized):
        return True
    return bool(re.search(r"(not ai|not generated|non ai|human|real|authentic|original)", normalized))


def _ai_score_from_labels(labels: list[dict[str, Any]]) -> tuple[str, float] | None:
    ai_label = next((entry for entry in labels if _is_ai_label(str(entry["label"]))), None)
    if ai_label:
        return str(ai_label["label"]), round_score(float(ai_label["score"]))
    human_label = next((entry for entry in labels if _is_human_label(str(entry["label"]))), None)
    if human_label:
        return str(human_label["label"]), round_score(1.0 - float(human_label["score"]))
    return None


def _classify_with_huggingface(input_data: str | bytes, model: str, settings: Settings) -> tuple[str, float]:
    if not settings.hf_token:
        raise RuntimeError("HF_TOKEN is not configured.")

    is_bytes = isinstance(input_data, bytes)
    payload = input_data if is_bytes else json.dumps({"inputs": input_data, "parameters": {"top_k": 2}}).encode("utf-8")
    request = urllib.request.Request(
        f"{settings.hf_inference_endpoint_base.rstrip('/')}/{model}",
        data=payload,
        headers={
            "Authorization": f"Bearer {settings.hf_token}",
            "Content-Type": "application/octet-stream" if is_bytes else "application/json",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(request, timeout=TIMEOUT_SECONDS) as response:
            labels = _flatten_response(json.loads(response.read().decode("utf-8")))
    except urllib.error.HTTPError as caught:
        raise RuntimeError(f"HF inference returned {caught.code}.") from caught

    mapped = _ai_score_from_labels(labels)
    if mapped is None:
        raise RuntimeError("HF inference returned no mappable AI/human label.")
    return mapped


def _classify_with_fallbacks(input_data: str | bytes, models: list[str], settings: Settings) -> tuple[str, str, float, list[dict[str, Any]]]:
    attempts: list[dict[str, Any]] = []
    for model in dict.fromkeys(models):
        try:
            label, score = _classify_with_huggingface(input_data, model, settings)
            attempts.append({"model": model, "status": "ok", "label": label, "score": score})
            return model, label, score, attempts
        except Exception as caught:
            attempts.append({"model": model, "status": "error", "error": str(caught)[:180]})
    raise RuntimeError(json.dumps({"message": "All HF model attempts failed.", "attempts": attempts}))


def _metadata(output: DetectorOutput) -> dict[str, Any]:
    metadata = output.artifacts.get("metadata")
    if isinstance(metadata, dict):
        return metadata
    output.artifacts["metadata"] = {}
    return output.artifacts["metadata"]


def _apply_hf_score(output: DetectorOutput, model: str, label: str, score: float, explanation: str) -> None:
    output.layer_scores["neural_classifier"] = score
    for layer in output.layer_breakdown:
        if layer.get("name") == "Neural Classifier":
            layer["score"] = score
            layer["explanation"] = explanation
    metadata = _metadata(output)
    metadata.update({"hf_model": model, "hf_label": label, "hf_score": score, "hf_status": "ok", "calibration_version": "hf-model-primary-v1"})


def _apply_hf_error(output: DetectorOutput, model: str, caught: Exception) -> None:
    metadata = _metadata(output)
    error = str(caught)[:180]
    attempts: Any = None
    try:
        attempts = json.loads(str(caught)).get("attempts")
    except Exception:
        attempts = [{"model": model, "status": "error", "error": error}]
    metadata.update({"hf_model": model, "hf_status": "fallback", "hf_error": error, "model_attempts": attempts, "calibration_version": "hf-model-primary-v1"})


def enrich_with_huggingface(output: DetectorOutput, settings: Settings, *, text: str | None = None, image: bytes | None = None) -> DetectorOutput:
    if not settings.hf_token:
        _metadata(output).update({"hf_status": "disabled"})
        return output

    if output.modality == "TEXT" and text:
        model = settings.hf_text_model_list[0]
        try:
            model, label, score, attempts = _classify_with_fallbacks(text, settings.hf_text_model_list, settings)
            _apply_hf_score(output, model, label, score, "Hugging Face Desklib AI text detector probability, with heuristic fallback available.")
            _metadata(output)["model_attempts"] = attempts
            output.explanation_parts.append("The neural classifier layer used Hugging Face Desklib AI text detector inference.")
        except Exception as caught:
            _apply_hf_error(output, model, caught)

    if output.modality == "IMAGE" and image:
        model = settings.hf_image_model_list[0]
        try:
            model, label, score, attempts = _classify_with_fallbacks(image, settings.hf_image_model_list, settings)
            _apply_hf_score(output, model, label, score, "Hugging Face AI image detector probability mapped from artificial-vs-real labels.")
            _metadata(output)["model_attempts"] = attempts
            output.explanation_parts.append("The neural classifier layer used Hugging Face AI image detector inference.")
            if score >= 0.58 and not any("Hugging Face" in source for source in output.detected_sources):
                output.detected_sources[:] = ["Hugging Face AI image detector", *output.detected_sources][:3]
        except Exception as caught:
            _apply_hf_error(output, model, caught)

    return output
